sql/schemachanger: add COMMENT ON TYPE operation to DSC

Dedej-Bergin opened the pull request:
We can now add comments to types in implicit transactions with the DSC (declarative schema changer) only.

Fixes: #123844
Release note (sql change): added COMMENT ON TYPE statement for implicit transactions

**scbuildstmt: reduced duplicate code in comment_on.go**

Added a dropComment and a addComment function to reuse code
and make the already existing functions more readable.

Release note: None

Dedej-Bergin committed:
sql/schemachanger: add COMMENT ON TYPE operation to DSC

We can now add comments to types in implicit transactions with
the DSC (declarative schema changer) only.

Fixes: #123844
Release note (sql change): added COMMENT ON TYPE statement
for implicit transactions

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124942)
<!-- Reviewable:end -->

spilchen reviewed (commented):
Looks good. I used this review to better understand the DSC, so some of my comments are simply questions due to my being new to this code area.

> Reviewed 46 of 46 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-:-O-H9bzE6Hhy78ma28sT:b-tx5kcd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @Dedej-Bergin, and @nkodali)

___
*[`pkg/sql/comment_on_type.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GqEIC5K11ykY3EMru:-O-GqEIC5K11ykY3EMrv:b-y32ofi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/comment_on_type.go#L23)):*
> ```Go
> // CommentOnType (UNIMPLEMENTED for legacy schema changer) add comment on a type.
> func (p *planner) CommentOnType(ctx context.Context, n *tree.CommentOnType) (planNode, error) {
> 	return nil, pgerror.New(pgcode.FeatureNotSupported, "COMMENT ON TYPE is only implemented in the declarative schema changer")
> ```

Do external customers know what the declarative schema changer is? Can we provide some more guidance about what they can do to resolve this? Maybe with a detail part to the message.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-Gv-DODi6E9aeKlFH8:-O-Gv-DODi6E9aeKlFH9:buwk0dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L115)):*
> ```Go
> 	typeElements := b.ResolveUserDefinedTypeType(n.Name, commentResolveParams)
> 	_, _, enumType := scpb.FindEnumType(typeElements)
> 	_, _, compositeType := scpb.FindCompositeType(typeElements)
> ```

We probably don't need to check if it it's a composite type if it's an enum. I suggest we make that conditional. We have conditional logic below to set the `typeID` and `typeObject` vars. I suggest we combine that into a single if/else branch.



_Suggestion:_
```Go
	var typeObject scpb.Element
	var typeID catid.DescID
	if _, _, enumType := scpb.FindEnumType(typeElements); enumType != nil {
		typeObject = enumType
		typeID = enumType.TypeID
	} else if _, _, compositeType := scpb.FindCompositeType(typeElements); compositeType != nil {
		typeObject = compositeType
		typeID = compositeType.TypeID
	} else {
		// error
	}
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GuboQAjy8tPvaD1gP:-O-GuboQAjy8tPvaD1gQ:blmbnm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L117)):*
> ```Go
> 	_, _, compositeType := scpb.FindCompositeType(typeElements)
> 
> 	typeName := n.Name.ToTableName()
> ```

It is odd that we convert the type name to a table name. Shouldn't `typeName` be defined as a `TypeName` rather than a `TableName` struct?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GwwM_-lb9Yvtgv81r:-O-GwwM_-lb9Yvtgv81s:b-igcdld) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L120)):*
> ```Go
> 
> 	if enumType == nil && compositeType == nil {
> 		b.MarkNameAsNonExistent(&typeName)
> ```

What is this code path needed for? If we try to alter an undefined type, don't we fail, via a panic, in `ResolveUserDefinedTypeType`
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-H1Z5x0cqut4kKqzdR:-O-H1Z5x0cqut4kKqzdS:b-ickn2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scpb/elements.proto#L150)):*
> ```proto
> 
>     // Type elements.
>      TypeComment type_comment = 180 [(gogoproto.moretags) = "parent:\"CompositeType,EnumType,AliasType\""];
> ```

nit: alignment seems off here. It right shifted by one extra character.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-H1gJAEpYEgYuIA8IN:-O-H1gJAEpYEgYuIA8IO:b-u0ramk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scpb/elements.proto#L150)):*
> ```proto
> 
>     // Type elements.
>      TypeComment type_comment = 180 [(gogoproto.moretags) = "parent:\"CompositeType,EnumType,AliasType\""];
> ```

Why is AliasType included here in the parent? Most of the new type code I saw only dealt with composite and enum types.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-H2B9F7HrG6U4kE_bP:-O-H2B9F7HrG6U4kE_bQ:bmzazvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scpb/elements.proto#L545)):*
> ```proto
> message TypeComment {
>  uint32 type_id = 1 [(gogoproto.customname) = "TypeID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
>  string comment = 2;
> ```

nit: alignment seems like its different from other messages in the file



_Suggestion:_
```proto
message TypeComment {
  uint32 type_id = 1 [(gogoproto.customname) = "TypeID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
  string comment = 2;
```
___
*[`pkg/util/log/eventpb/ddl_events.proto` line 254 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-H6ark2GUlfIbQFb9i:-O-H6arlAVtzuM6729h0:bxg62uc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/util/log/eventpb/ddl_events.proto#L254)):*
> ```proto
> }
> 
> // CommentOnType is recorded when a table is commented.
> ```

nit: copy/paste error?



_Suggestion:_
```proto
// CommentOnType is recorded when a type is commented.
```
___
*[`pkg/util/log/eventpb/ddl_events.proto` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-H6RYd3rbS08Sv8ilI:-O-H6RYd3rbS08Sv8ilJ:b-3zp7f2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/util/log/eventpb/ddl_events.proto#L258)):*
> ```proto
>   CommonEventDetails common = 1 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "", (gogoproto.embed) = true];
>   CommonSQLEventDetails sql = 2 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "", (gogoproto.embed) = true];
>   // The name of the affected table.
> ```

nit: minor correction in the comment



_Suggestion:_
```proto
// The name of the affected type.
```
___
*[`pkg/sql/logictest/testdata/logic_test/comment_on` line 0 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-Gr6VXBziXVYN7BOWe:-O-Gr6VXBziXVYN7BOWf:b-shmb6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/logictest/testdata/logic_test/comment_on#L0)):*
I was running on your branch, and I noticed that the the `\dT` command fails if there is a comment attached to the type.
```
demo@127.0.0.1:26257/demoapp/movr> CREATE TYPE roach_dwellings AS ENUM
                                -> ('roach_motel','roach_kitchen','roach_bathroom','roach_house');
CREATE TYPE

Time: 8ms total (execution 7ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr> \dT;
List of data types:
  Schema |      Name       | Description
---------+-----------------+--------------
  public | roach_dwellings |
(1 row)
demo@127.0.0.1:26257/demoapp/movr> COMMENT ON TYPE roach_dwellings IS 'test';
COMMENT ON TYPE

Time: 10ms total (execution 9ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr> \dT;
List of data types:
ERROR: internal error: column kv_catalog_comments.classoid not nullable, but found NULL value
SQLSTATE: XX000
```
___
*[`pkg/sql/logictest/testdata/logic_test/comment_on` line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GqmE6Egdf8WaNn62-:-O-GqmE72dE-1Dky2lVd:bbwnr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/logictest/testdata/logic_test/comment_on#L287)):*
> ```
> skipif config local-mixed-23.2
> statement ok
> COMMENT ON TYPE roach_dwellings IS 'First-CRDB-comment-on-types';
> ```

If it's not tested elsewhere, can we verify that we fail if running the legacy schema changer?
___
*[`pkg/sql/schemachanger/scplan/testdata/comment_on` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-Gbn3X7s9OF0TVdTgq:-O-Gbn3X7s9OF0TVdTgr:blts0zt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scplan/testdata/comment_on#L391)):*
> ```
> 
> ops
> COMMENT ON TYPE roach_legs IS NULL;
> ```

Should we test that a comment can be changed to some other value that isn't NULL? i.e.
```
COMMENT ON TYPE roach_legs IS 'Updated existing comment';
```
___
*[`pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition` line 0 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GjVS2BsZJUIXpwIW7:-O-GjVS2BsZJUIXpwIW8:b-eipros) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition#L0)):*
Should the directory and file name be `comment_on_type` to allow for future tests of other 'COMMENT ON' commands?
___
*[`pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GhNxB4oTPctFq0v4A:-O-GhNxC1sb8DLIEUoOT:bqc6cks) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition#L6)):*
> ```
> 
> test
> COMMENT ON TYPE roach_dwellings IS 'First CRDB comment on types';
> ```

Is `COMMENT ON TYPE` sufficiently different for enum and composite types that it makes sense to test both?


<!-- Sent from Reviewable.io -->

Dedej-Bergin committed:
scbuildstmt: reduced duplicate code in comment_on.go

Added a dropComment and a addComment function to reuse code
and make the already existing functions more readable.

Release note: None

Dedej-Bergin commented:
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-Gv-DODi6E9aeKlFH8:-O-Hck1k-soR_V-EdbBZ:b-ntoya6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L115)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

We probably don't need to check if it it's a composite type if it's an enum. I suggest we make that conditional. We have conditional logic below to set the `typeID` and `typeObject` vars. I suggest we combine that into a single if/else branch.
</blockquote></details>

Sounds good thanks! I just removed this in my second commit which contains a refactoring of comment_on.go.


<!-- Sent from Reviewable.io -->

Dedej-Bergin commented:
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GuboQAjy8tPvaD1gP:-O-Hd2qI0XRvEKUae13u:b-wctvu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L117)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

It is odd that we convert the type name to a table name. Shouldn't `typeName` be defined as a `TypeName` rather than a `TableName` struct?
</blockquote></details>

Yes this is odd, thanks! I just removed this in my second commit which contains a refactoring of comment_on.go.


<!-- Sent from Reviewable.io -->

Dedej-Bergin commented:
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GwwM_-lb9Yvtgv81r:-O-HdJNO4hFx0WPysLz5:bbbx6mx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L120)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

What is this code path needed for? If we try to alter an undefined type, don't we fail, via a panic, in `ResolveUserDefinedTypeType`
</blockquote></details>

Yes it's not needed thanks! I just removed this in my second commit which contains a refactoring of comment_on.go.


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @nkodali, and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-Gv-DODi6E9aeKlFH8:-O-I4pfxA1R1x5tfsw7g:b76rs3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L115)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Sounds good thanks! I just removed this in my second commit which contains a refactoring of comment_on.go.
</blockquote></details>

I flipped the commits around, as it works better for addressing/amending code review changes.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GuboQAjy8tPvaD1gP:-O-I4Bj72cImkWlU_AEa:b76rs3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L117)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Yes this is odd, thanks! I just removed this in my second commit which contains a refactoring of comment_on.go.
</blockquote></details>

I flipped the commits around, as it works better for addressing/amending code review changes.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GwwM_-lb9Yvtgv81r:-O-I4qvL1qG3NhK72gug:b76rs3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L120)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Yes it's not needed thanks! I just removed this in my second commit which contains a refactoring of comment_on.go.
</blockquote></details>

I flipped the commits around, as it works better for addressing/amending code review changes.


<!-- Sent from Reviewable.io -->

Dedej-Bergin commented:
*[`pkg/sql/comment_on_type.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GqEIC5K11ykY3EMru:-O-IA5CoEpp09fZ2SLDg:bwbzb54) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/comment_on_type.go#L23)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Do external customers know what the declarative schema changer is? Can we provide some more guidance about what they can do to resolve this? Maybe with a detail part to the message.
</blockquote></details>

Good point, most external customers are not aware of the declarative schema changer.  I removed mentioning declarative schema changer and called it implicit transactions instead. Thanks Matt!


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
I like the refactor. It is much simpler.

> Reviewed 5 of 6 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-:-O-MB7oq2nSMOuUrHROl:bwu02i8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @Dedej-Bergin, and @nkodali)

___
*[`pkg/sql/comment_on_type.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GqEIC5K11ykY3EMru:-O-M3s_29s-9N4N1m8UR:bqdd0q6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/comment_on_type.go#L23)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Good point, most external customers are not aware of the declarative schema changer.  I removed mentioning declarative schema changer and called it implicit transactions instead. Thanks Matt!
</blockquote></details>

Sorry, I never search the docs before making this comment. But it does look like we document the declarative schema changer. See https://www.cockroachlabs.com/docs/v24.1/online-schema-changes.html#declarative-schema-changer. So, the original message was fine. I would still add a hint to let the user know how to enable the declarative schema changer. Maybe something like:
```
		return errors.WithHint(pgerror.Newf(pgcode.FeatureNotSupported,
			"COMMENT ON TYPE is only implemented in the declarative schema changer"),
			`Use the "use_declarative_schema_changer" session variable or "sql.defaults.use_declarative_schema_changer"`+
				`cluster setting to enable the declarative schema changer.`)
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 77 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-MAzEq6tau1AsyQMv2:-O-MAzEq6tau1AsyQMv3:b-nqn4k7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52c53352e111ced4b1135334e21809b52a84940/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L77)):*
> ```Go
> 	typeElements := b.ResolveUserDefinedTypeType(n.Name, commentResolveParams)
> 
> 	if n.Comment == nil {
> ```

Much simpler function. I like it.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 165 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-M9yo28SA5DKo4R5kb:-O-M9yo28SA5DKo4R5kc:bxfqq59) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52c53352e111ced4b1135334e21809b52a84940/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L165)):*
> ```Go
> 				comment.(*scpb.ColumnComment).ColumnID = object.ColumnID
> 				comment.(*scpb.ColumnComment).PgAttributeNum = object.PgAttributeNum
> 			}
> ```

Nice refactor! Can we add a default case that will panic? That will serve as a reminder to update this if there is another thing we need to comment on.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 179 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-MAJuz52BvHHUjMlP-:-O-MAJuz52BvHHUjMlP0:bti3fu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52c53352e111ced4b1135334e21809b52a84940/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L179)):*
> ```Go
> 			comment.(*scpb.IndexComment).IndexID = object.IndexID
> 		case *scpb.SecondaryIndex:
> 			comment.(*scpb.IndexComment).IndexID = object.IndexID
> ```

nit: each assignment line seems to be the same. Any reason we couldn't combine them?



_Suggestion:_
```Go
		case *scpb.PrimaryIndex, *scpb.SecondaryIndex:
			comment.(*scpb.IndexComment).IndexID = object.IndexID
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-MAYZyCxcKE5KmCPBI:-O-MAYZyCxcKE5KmCPBJ:b-rmm03e) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52c53352e111ced4b1135334e21809b52a84940/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L187)):*
> ```Go
> 	if reflect.TypeOf(comment) == reflect.TypeOf(&scpb.ConstraintComment{}) {
> 		switch object := e.(type) {
> 		case *scpb.PrimaryIndex:
> ```

nit: similar comment here about duplicating the assignment line. Maybe we can combine them.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @Dedej-Bergin, @nkodali, and @spilchen)

___
*[`pkg/sql/comment_on_type.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GqEIC5K11ykY3EMru:-O-MbNuf3iU9-G7IeDlh:b-tdfk21) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/comment_on_type.go#L23)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Sorry, I never search the docs before making this comment. But it does look like we document the declarative schema changer. See https://www.cockroachlabs.com/docs/v24.1/online-schema-changes.html#declarative-schema-changer. So, the original message was fine. I would still add a hint to let the user know how to enable the declarative schema changer. Maybe something like:
```
		return errors.WithHint(pgerror.Newf(pgcode.FeatureNotSupported,
			"COMMENT ON TYPE is only implemented in the declarative schema changer"),
			`Use the "use_declarative_schema_changer" session variable or "sql.defaults.use_declarative_schema_changer"`+
				`cluster setting to enable the declarative schema changer.`)
```
</blockquote></details>

One thing to watch out for is that the declarative schema changer already is on by default at this point, but we can still fallback to using the legacy schema changer if a schema change statement is used inside of an explicit (or multi-statement) transaction.

I think it makes sense to keep the message "COMMENT ON TYPE is only implemented in the declarative schema changer" but without the extra hint. Instead, we can add a hint whenever we fallback from the declarative to the legacy schema changer (in a separate PR).
___
*[`pkg/sql/logictest/testdata/logic_test/comment_on` line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GqmE6Egdf8WaNn62-:-O-MdM6v4Cvb29z2vePv:b-e5tqe2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/logictest/testdata/logic_test/comment_on#L287)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

If it's not tested elsewhere, can we verify that we fail if running the legacy schema changer?
</blockquote></details>

+1. the way we can test this is with the directive: `onlyif config local-legacy-schema-changer`
___
*[`pkg/sql/logictest/testdata/logic_test/comment_on` line  at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-Gr6VXBziXVYN7BOWe:-O-MdqKR0JnW_z1uw02X:b-wy1ta4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/logictest/testdata/logic_test/comment_on#L)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I was running on your branch, and I noticed that the the `\dT` command fails if there is a comment attached to the type.
```
demo@127.0.0.1:26257/demoapp/movr> CREATE TYPE roach_dwellings AS ENUM
                                -> ('roach_motel','roach_kitchen','roach_bathroom','roach_house');
CREATE TYPE

Time: 8ms total (execution 7ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr> \dT;
List of data types:
  Schema |      Name       | Description
---------+-----------------+--------------
  public | roach_dwellings |
(1 row)
demo@127.0.0.1:26257/demoapp/movr> COMMENT ON TYPE roach_dwellings IS 'test';
COMMENT ON TYPE

Time: 10ms total (execution 9ms / network 0ms)

demo@127.0.0.1:26257/demoapp/movr> \dT;
List of data types:
ERROR: internal error: column kv_catalog_comments.classoid not nullable, but found NULL value
SQLSTATE: XX000
```
</blockquote></details>

nice find. let's make sure to get the underlying query used by `\dT` tested here in the logic tests. it is:
```
   SELECT n.nspname AS "Schema",
          pg_catalog.format_type(t.oid, NULL) AS "Name",
          COALESCE(pg_catalog.obj_description(t.oid, 'pg_type'),'') AS "Description"
     FROM pg_catalog.pg_type t
LEFT JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
    WHERE (t.typrelid = 0
          OR (SELECT c.relkind = 'c'
                FROM pg_catalog.pg_class c
               WHERE c.oid = t.typrelid))
      AND (NOT EXISTS(
          SELECT 1
            FROM pg_catalog.pg_type el
           WHERE el.oid = t.typelem AND el.typarray = t.oid))
      AND n.nspname !~ '^pg_'
      AND n.nspname <> 'information_schema'
      AND n.nspname <> 'crdb_internal'
      AND pg_catalog.pg_type_is_visible(t.oid)
ORDER BY 1, 2
```

The problem here is likely with `obj_description`, so let's make sure to add tests for that specifically too. We can add that to the `builtin_function` testdata file, where there are some other tests for that function.
___
*[`pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GhNxB4oTPctFq0v4A:-O-MdWSyCq1dpE7rOzpa:b7rb4d5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition#L6)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Is `COMMENT ON TYPE` sufficiently different for enum and composite types that it makes sense to test both?
</blockquote></details>

yeah, we have been frequently bitten by slight non-obvious differences between the two types. let's test both.

also, can you add an end_to_end test for removing the comment also?


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @nkodali, and @spilchen)

___
*[`pkg/sql/comment_on_type.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GqEIC5K11ykY3EMru:-O-aZRhz5W96Ybt-PPyN:b-ne7bvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/comment_on_type.go#L23)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

One thing to watch out for is that the declarative schema changer already is on by default at this point, but we can still fallback to using the legacy schema changer if a schema change statement is used inside of an explicit (or multi-statement) transaction.

I think it makes sense to keep the message "COMMENT ON TYPE is only implemented in the declarative schema changer" but without the extra hint. Instead, we can add a hint whenever we fallback from the declarative to the legacy schema changer (in a separate PR).
</blockquote></details>

Sounds good, I changed it back.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 77 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-MAzEq6tau1AsyQMv2:-O-aZWI_6oaqkgBV6ZLa:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52c53352e111ced4b1135334e21809b52a84940/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L77)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Much simpler function. I like it.
</blockquote></details>

Thanks!
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 165 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-M9yo28SA5DKo4R5kb:-O-aZy-r5-USuY9db_ue:b-o2cybp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52c53352e111ced4b1135334e21809b52a84940/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L165)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Nice refactor! Can we add a default case that will panic? That will serve as a reminder to update this if there is another thing we need to comment on.
</blockquote></details>

Thanks for the suggestion! I don't think we can add it though.

I tried adding the panic and it ends up causing a few test failures.  I think we're not able to add a default case for the panic because the elements result set is expected to have other types then the ones caught by the cases.  When an element is not caught by a case then we want to just do nothing.

Here's what I tried and one of the test failures



_Code snippet:_
```Go
case \*scpb.Column:  
  if reflect.TypeOf(comment) == reflect.TypeOf(&scpb.ColumnComment{}) {  
   comment.(\*scpb.ColumnComment).ColumnID = object.ColumnID  
   comment.(\*scpb.ColumnComment).PgAttributeNum = object.PgAttributeNum  
  }  
default:  
  panic(pgerror.Newf(pgcode.Syntax, "unexpected element type %T", e))  
}

// One of the testcase failures
github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/pg_catalog:2719: 
        expected success, but found
        (42601) unexpected element type *scpb.Namespace
        comment_on.go:171: in func1()
    panic.go:626: -- test log scope end --
FAIL
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 179 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-MAJuz52BvHHUjMlP-:-O-a_EvY3Gzl1_VRpvxa:b-us52di) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52c53352e111ced4b1135334e21809b52a84940/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L179)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: each assignment line seems to be the same. Any reason we couldn't combine them?
</blockquote></details>

No reason, I went ahead and made this change. Thanks Matt!
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-MAYZyCxcKE5KmCPBI:-O-a_IkM8nwwqZwgZIrh:bl06c80) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52c53352e111ced4b1135334e21809b52a84940/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L187)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: similar comment here about duplicating the assignment line. Maybe we can combine them.
</blockquote></details>

Yes, definitely looks better combined. I just made this change, thanks Matt!
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-H1Z5x0cqut4kKqzdR:-O-a_mD60qLMdhrK6ZxD:b-vrgo1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scpb/elements.proto#L150)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: alignment seems off here. It right shifted by one extra character.
</blockquote></details>

Thanks for catching this, just fixed it.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-H2B9F7HrG6U4kE_bP:-O-aa3UyBqXvJprJbGKL:b-vrgo1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scpb/elements.proto#L545)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: alignment seems like its different from other messages in the file
</blockquote></details>

Thanks for catching this, just fixed it.
___
*[`pkg/util/log/eventpb/ddl_events.proto` line 254 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-H6ark2GUlfIbQFb9i:-O-aa4nf2FCGsvZEcrMu:b-vrgo1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/util/log/eventpb/ddl_events.proto#L254)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: copy/paste error?
</blockquote></details>

Thanks for catching this, just fixed it.
___
*[`pkg/util/log/eventpb/ddl_events.proto` line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-H6RYd3rbS08Sv8ilI:-O-aaLAA51jX03WHGsHp:b-vrgo1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/util/log/eventpb/ddl_events.proto#L258)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: minor correction in the comment
</blockquote></details>

Thanks for catching this, just fixed it.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> Reviewed 1 of 2 files at r4, 4 of 4 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @Dedej-Bergin, and @nkodali)

___
*[`pkg/sql/comment_on_type.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GqEIC5K11ykY3EMru:-O-gL2BzBG6OmUUu5UrK:b-dxcbts) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/comment_on_type.go#L23)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Sounds good, I changed it back.
</blockquote></details>

Makes sense.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go` line 165 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-M9yo28SA5DKo4R5kb:-O-gNWYZCdwc4AAguV0g:bkdg53) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52c53352e111ced4b1135334e21809b52a84940/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/comment_on.go#L165)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Thanks for the suggestion! I don't think we can add it though.

I tried adding the panic and it ends up causing a few test failures.  I think we're not able to add a default case for the panic because the elements result set is expected to have other types then the ones caught by the cases.  When an element is not caught by a case then we want to just do nothing.

Here's what I tried and one of the test failures
</blockquote></details>

Okay, thanks for trying to add it.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
Another thing I just noticed is that the `SHOW CREATE ALL..` commands include DDL for any comments added. See this: https://www.cockroachlabs.com/docs/v24.1/show-create.html#show-the-create-table-statement-for-a-table-with-a-comment

Do we need to do this for `SHOW CREATE ALL TYPES` as well? If it's required, we may want to handle it in a separate PR since this one is already quite large.

> Reviewed 10 of 10 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-:-O-gpKRC9wVHuaNjrmkz:b-1oesxz)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @Dedej-Bergin, and @nkodali)



<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @nkodali, and @spilchen)

___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-H1gJAEpYEgYuIA8IN:-O-g-mVn1-jbgbuUlufq:b-icvhqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scpb/elements.proto#L150)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Why is AliasType included here in the parent? Most of the new type code I saw only dealt with composite and enum types.
</blockquote></details>

It shouldn't be there, I went ahead and removed it. Thanks Matt!
___
*[`pkg/sql/logictest/testdata/logic_test/comment_on` line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GqmE6Egdf8WaNn62-:-O-ftJcLCOIl3HuHfk8z:b3l2ghs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/logictest/testdata/logic_test/comment_on#L287)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

+1. the way we can test this is with the directive: `onlyif config local-legacy-schema-changer`
</blockquote></details>

Thanks, just added this test case.
___
*[`pkg/sql/logictest/testdata/logic_test/comment_on` line  at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-Gr6VXBziXVYN7BOWe:-O-flrrl4SINzluyY2S3:bxg5bzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/logictest/testdata/logic_test/comment_on#L)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nice find. let's make sure to get the underlying query used by `\dT` tested here in the logic tests. it is:
```
   SELECT n.nspname AS "Schema",
          pg_catalog.format_type(t.oid, NULL) AS "Name",
          COALESCE(pg_catalog.obj_description(t.oid, 'pg_type'),'') AS "Description"
     FROM pg_catalog.pg_type t
LEFT JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
    WHERE (t.typrelid = 0
          OR (SELECT c.relkind = 'c'
                FROM pg_catalog.pg_class c
               WHERE c.oid = t.typrelid))
      AND (NOT EXISTS(
          SELECT 1
            FROM pg_catalog.pg_type el
           WHERE el.oid = t.typelem AND el.typarray = t.oid))
      AND n.nspname !~ '^pg_'
      AND n.nspname <> 'information_schema'
      AND n.nspname <> 'crdb_internal'
      AND pg_catalog.pg_type_is_visible(t.oid)
ORDER BY 1, 2
```

The problem here is likely with `obj_description`, so let's make sure to add tests for that specifically too. We can add that to the `builtin_function` testdata file, where there are some other tests for that function.
</blockquote></details>

Thanks guys! Fixed this by adding the below 3 lines to crdb_internal.go at line 5939, and I also added the test for the underlying query use by \DT.



_Code snippet:_
```
case catalogkeys.IndexCommentType:  
  classOid = tree.NewDOid(catconstants.PgCatalogClassTableID)  
  objOid = h.IndexOid(descpb.ID(key.ObjectID), descpb.IndexID(key.SubID))
```
___
*[`pkg/sql/schemachanger/scplan/testdata/comment_on` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-Gbn3X7s9OF0TVdTgq:-O-ftRJuA85V_DzyKLT6:blnm6qp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scplan/testdata/comment_on#L391)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Should we test that a comment can be changed to some other value that isn't NULL? i.e.
```
COMMENT ON TYPE roach_legs IS 'Updated existing comment';
```
</blockquote></details>

Yes we should, just added this, thanks Matt!
___
*[`pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GhNxB4oTPctFq0v4A:-O-g2SEc1hxhWut5DoU0:b-vc8426) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition#L6)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

yeah, we have beeb frequently bitten by slight non-obvious differences between the two types. let's test both.

also, can you add an end_to_end test for removing the comment also?
</blockquote></details>

Sounds good, I went ahead and added the test for removing the comment, thanks!
___
*[`pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition` line  at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GjVS2BsZJUIXpwIW7:-O-fmsgzCMJGRWtE5cfi:b-jkrgiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition#L)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Should the directory and file name be `comment_on_type` to allow for future tests of other 'COMMENT ON' commands?
</blockquote></details>

Yes, I went ahead and made this change. Thanks Matt!


<!-- Sent from Reviewable.io -->

Dedej-Bergin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @nkodali, and @spilchen)

___
*[`pkg/sql/schemachanger/scplan/testdata/comment_on` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-Gbn3X7s9OF0TVdTgq:-O-hST645cLcIOQRSQ6F:b61pfkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/scplan/testdata/comment_on#L391)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Yes we should, just added this, thanks Matt!
</blockquote></details>

I added this to /pkg/sql/logictest/testdata/logic_test/comment_on


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> Reviewed 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @Dedej-Bergin, @nkodali, and @rafiss)

___
*[`pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GhNxB4oTPctFq0v4A:-O-kgCMh23l2Gy1OU5WF:b-ucjb64) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition#L6)):*
<details><summary><i>Previously, Dedej-Bergin (Bergin Dedej) wrote…</i></summary><blockquote>

Sounds good, I went ahead and added the test for removing the comment, thanks!
</blockquote></details>

Maybe I am missing something, but I didn't see where we test composite types in this `end_to_end` test suite. 


<!-- Sent from Reviewable.io -->

Dedej-Bergin commented:
*[`pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-GhNxB4oTPctFq0v4A:-O-lhh-u7Ac3h-tYGB-o:b-wyr63g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0f0c4e6dbc6a3a1bb58a9df825336e1171189ce/pkg/sql/schemachanger/testdata/end_to_end/comment_on/comment_on.definition#L6)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Maybe I am missing something, but I didn't see where we test composite types in this `end_to_end` test suite.
</blockquote></details>

I misread this for some reason, I just added the tests now. Thanks for clarifying Matt!


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> Reviewed 19 of 19 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @Dedej-Bergin, @nkodali, and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 4578 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-m3GGb7l7Z_QdFLhZF:-O-m3GGb7l7Z_QdFLhZG:b-fwe1qp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e2e81169274e169ec842b865f452c30092876c6/pkg/sql/crdb_internal.go#L4578)):*
> ```Go
> 			return nil, nil, err
> 		}
> 		viewActOrViewActRedact, _, err := p.HasViewActivityOrViewActivityRedactedRole(ctx)
> ```

This looks like Faizan's change that recently went into master. It's weird to see it in this review. Is this just a quirk of the diff tool?


<!-- Sent from Reviewable.io -->

Dedej-Bergin commented:
*[`pkg/sql/crdb_internal.go` line 4578 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-O-m3GGb7l7Z_QdFLhZF:-O-m4J241kXWMIF6BYg0:b9itbjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e2e81169274e169ec842b865f452c30092876c6/pkg/sql/crdb_internal.go#L4578)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

This looks like Faizan's change that recently went into master. It's weird to see it in this review. Is this just a quirk of the diff tool?
</blockquote></details>

I think it's a quirk of the diff tool, the github diff doesn't show it.
![Screenshot 2024-06-19 at 3.59.47 PM.png](https://files.reviewableusercontent.io/images/reviewable/164547647/NipPNSgEGPmZ1HTzIvXsqd6d/Screenshot%202024-06-19%20at%203.59.47%E2%80%AFPM.png)


<!-- Sent from Reviewable.io -->

spilchen reviewed (approved):
lgtm

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-:-O-m771x7O8ZdpJj_6O2:bb7k0qh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom, @nkodali, and @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124942#-:-O-qDb8b6VY42UCthT4h:b-7nbn20)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @nkodali)



<!-- Sent from Reviewable.io -->

Dedej-Bergin commented:
bors r+
